src/app/clusters/scenes-server/scenes-server.cpp

```diff
@@ -0,0 +1,993 @@
+        HandleCommand<Commands::StoreScene::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleStoreScene(ctx, req); });
+        return;
+    case Commands::RecallScene::Id:
+        HandleCommand<Commands::RecallScene::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleRecallScene(ctx, req); });
+        return;
+    case Commands::GetSceneMembership::Id:
+        HandleCommand<Commands::GetSceneMembership::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleGetSceneMembership(ctx, req); });
+        return;
+    case Commands::EnhancedAddScene::Id:
+        HandleCommand<Commands::EnhancedAddScene::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleEnhancedAddScene(ctx, req); });
+        return;
+    case Commands::EnhancedViewScene::Id:
+        HandleCommand<Commands::EnhancedViewScene::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleEnhancedViewScene(ctx, req); });
+        return;
+    case Commands::CopyScene::Id:
+        HandleCommand<Commands::CopyScene::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleCopyScene(ctx, req); });
+        return;
+    }
+}
+
+// AttributeAccessInterface
+CHIP_ERROR ScenesServer::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    uint8_t value             = 0;
+    uint16_t endpoinTableSize = 0;
+    ReturnErrorOnFailure(
+        StatusIB(ToInteractionModelStatus(Attributes::SceneTableSize::Get(aPath.mEndpointId, &endpoinTableSize))).ToChipError());
+
+    // Get Scene Table Instance
+    scenes::SceneTable<scenes::ExtensionFieldSetsImpl> * sceneTable;
+
+    switch (aPath.mAttributeId)
+    {
+    case Attributes::SceneCount::Id:
+        sceneTable = chip::scenes::GetSceneTableImpl(aPath.mEndpointId, endpoinTableSize);
+        ReturnErrorOnFailure(sceneTable->GetEndpointSceneCount(value));
+        return aEncoder.Encode(value);
+    case Attributes::RemainingCapacity::Id:
+        sceneTable = chip::scenes::GetSceneTableImpl(aPath.mEndpointId, endpoinTableSize);
+        ReturnErrorOnFailure(sceneTable->GetRemainingCapacity(aEncoder.AccessingFabricIndex(), value));
+        return aEncoder.Encode(value);
+    default:
+        return CHIP_NO_ERROR;
+    }
+}
+
+void ScenesServer::GroupWillBeRemoved(FabricIndex aFabricIx, EndpointId aEndpointId, GroupId aGroupId)
+{
+    uint16_t endpoinTableSize = 0;
+    VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == Attributes::SceneTableSize::Get(aEndpointId, &endpoinTableSize));
+
+    // Get Scene Table Instance
+    scenes::SceneTable<scenes::ExtensionFieldSetsImpl> * sceneTable =
+        chip::scenes::GetSceneTableImpl(aEndpointId, endpoinTableSize);
+    VerifyOrReturn(nullptr != sceneTable);
+
+    VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == Attributes::SceneValid::Set(aEndpointId, false));
```

Comment:
So if this fails we will leak the scenes in storage?  I think we should ignore failure from this `Set()` call (and document why) to make sure we clean the storage up.  Maybe just call MakeSceneInvalid() here?

That said, it's not obvious why we would mark SceneValid as false unless the actual scene we were in is deleted.  Followup to get the logic here sorted out better, please.